### PR TITLE
AFL++ Qemu mode

### PR DESCRIPTION
**Pull request by andreafioraldi**
let's try QEMU mode in fuzzbench, the chosen configuration is floating-point CompCov + persistent mode

### CODE REVIEW

**Comment by andreafioraldi** [comment]
@googlebot I consent.

File: .github/workflows/fuzzers.yml [hunk 1]
```diff
@@ -19,6 +19,7 @@ jobs:
           - aflplusplus
           - aflplusplus_optimal
           - aflplusplus_shmem
+          - aflplusplus_optimal_shmem
```
Comment (jonathanmetzman) [inline .github/workflows/fuzzers.yml:22]: I'm confused by this? Isn't this already in fuzzers.yml?

Comment (jonathanmetzman) [inline .github/workflows/fuzzers.yml:22]: Shouldn't we add aflplusplus_qemu here?

Comment (andreafioraldi) [reply .github/workflows/fuzzers.yml:22]: yes I forgot it

Comment (andreafioraldi) [reply .github/workflows/fuzzers.yml:22]: @vanhauser-thc

**Comment by jonathanmetzman** [comment]
yay!

Please see presubmit errors btw: https://github.com/google/fuzzbench/pull/396/checks?check_run_id=734199564

need to run `make format`
and fix some issues reported by `make lint`.

**Comment by andreafioraldi** [comment]
> yay!
> 
> Please see presubmit errors btw: https://github.com/google/fuzzbench/pull/396/checks?check_run_id=734199564
> 
> need to run `make format`
> and fix some issues reported by `make lint`.

yes, seems that the code is not checked if I do git add before presubmit, should be fine ok.

**Comment by jonathanmetzman** [comment]
> > yay!
> > Please see presubmit errors btw: https://github.com/google/fuzzbench/pull/396/checks?check_run_id=734199564
> > need to run `make format`
> > and fix some issues reported by `make lint`.
> 
> yes, seems that the code is not checked if I do git add before presubmit, should be fine now.

You're saying if you do `git add file.py` and then `make presubmit` errors in `file.py` aren't detected? I wasn't able to repro this when I just tried, but I've seen our git stuff not work with different set ups (maybe related to using a fork of the repo?).

**Comment by vanhauser-thc** [comment]
this will need another change (a new git checkout) stay tuned :)

**Comment by vanhauser-thc** [comment]
the build system seems to be very brittle at the moment. 
Several - but curiously not all - builds were failing with:
```
+ cp BUILD/libexpat/expat/lib/.libs/libexpat.so.1.6.11 /out/
+ ln -s /out/libexpat.so.1.6.11 /out/libexpat.so.1
+ ln -s /out/libexpat.so.1.6.11 /out/libexpat.so
+ cp BUILD/libdbus/dbus/.libs/libdbus-1.so.3.14.16 /out/
cp: cannot stat 'BUILD/libdbus/dbus/.libs/libdbus-1.so.3.14.16': No such file or directory
```
this seems to be an issue in fuzzbench or in travis networking.

then in aflplusplus_optimal_shmem the systemd built errors while compiling without any error output, but it works fine: 
```
make run-aflplusplus_optimal-systemd_fuzz-link-parser
...
[*] Entering queue cycle 1.
[*] Fuzzing test case #0 (6 total, 0 uniq crashes found)...
[*] Fuzzing test case #1 (157 total, 0 uniq crashes found)...
```

In a previous run I had the same thing with freetype:
```
make run-aflplusplus_shmem-freetype2-2017
...
[*] Fuzzing test case #5416 (5649 total, 0 uniq crashes found)...
[*] Fuzzing test case #5420 (5649 total, 0 uniq crashes found)...
[*] Fuzzing test case #5422 (5649 total, 0 uniq crashes found)...
[*] Fuzzing test case #5434 (5653 total, 0 uniq crashes found)...
[*] Fuzzing test case #5435 (5660 total, 0 uniq crashes found)...
```

**Comment by jonathanmetzman** [comment]
> the build system seems to be very brittle at the moment.
> Several - but curiously not all - builds were failing with:
> 
> ```
> + cp BUILD/libexpat/expat/lib/.libs/libexpat.so.1.6.11 /out/
> + ln -s /out/libexpat.so.1.6.11 /out/libexpat.so.1
> + ln -s /out/libexpat.so.1.6.11 /out/libexpat.so
> + cp BUILD/libdbus/dbus/.libs/libdbus-1.so.3.14.16 /out/
> cp: cannot stat 'BUILD/libdbus/dbus/.libs/libdbus-1.so.3.14.16': No such file or directory
> ```
> 
> this seems to be an issue in fuzzbench or in travis networking.
> 
> then in aflplusplus_optimal_shmem the systemd built errors while compiling without any error output, but it works fine:
> 
> ```
> make run-aflplusplus_optimal-systemd_fuzz-link-parser
> ...
> [*] Entering queue cycle 1.
> [*] Fuzzing test case #0 (6 total, 0 uniq crashes found)...
> [*] Fuzzing test case #1 (157 total, 0 uniq crashes found)...
> ```
> 
> In a previous run I had the same thing with freetype:
> 
> ```
> make run-aflplusplus_shmem-freetype2-2017
> ...
> [*] Fuzzing test case #5416 (5649 total, 0 uniq crashes found)...
> [*] Fuzzing test case #5420 (5649 total, 0 uniq crashes found)...
> [*] Fuzzing test case #5422 (5649 total, 0 uniq crashes found)...
> [*] Fuzzing test case #5434 (5653 total, 0 uniq crashes found)...
> [*] Fuzzing test case #5435 (5660 total, 0 uniq crashes found)...
> ```

Should I restart the builds? The AFL+AFL++ builds for OSS-Fuzz passed and the standard onea look like they will pass. Based  on this I wouldn't think it's a network error.

**Comment by vanhauser-thc** [comment]
yes please restart those which failed.
as you can see also `afl` failed - and obviously we did not touch that :)

**Comment by jonathanmetzman** [comment]
I was able to repro the failure in TCP dump so I'm going to remove it #400. 
You can pull from master and push again if you want to test, or I can land.
But I didn't really follow what you were saying about systemd. And in the OSS-Fuzz benchmarks, it looks like qemu, optimal shmem, and shmem fail on runtime on jsoncpp benchmark.

**Comment by jonathanmetzman** [comment]
Looks like only issue is a qemu forkserver problem with jsoncpp

**Comment by vanhauser-thc** [comment]
hmmm I only changed the builder.Dockerfile of qemu ... though all aflplusplus variants seem to get a full build test. I thought that changed?

**Comment by jonathanmetzman** [comment]
> hmmm I only changed the builder.Dockerfile of qemu ... though all aflplusplus variants seem to get a full build test. I thought that changed?

That happens on a per-pull request basis, so even if you didn't modify it during the last commit, it is built because it was modified earlier in the PR.

**Comment by jonathanmetzman** [comment]
I'll try to do an experiment with the changed fuzzers later today or Saturday night.

**Comment by vanhauser-thc** [comment]
put honggfuzz in there please ;)

**Comment by jonathanmetzman** [comment]
> put honggfuzz in there please ;)

Are you fine with me reusing the 5/24 one or do you want a fresh one?

**Comment by vanhauser-thc** [comment]
if it is still the same commit id then the original should be fine. thanks for asking!

**Comment by jonathanmetzman** [comment]
> if it is still the same commit id then the original should be fine. thanks for asking!

It is.

**Comment by vanhauser-thc** [comment]
I just noticed that for the shmem variants that they dont process input if the target is not given as @@ (so no stdin) - is that an issue for fuzzbench?
 I have never checked how the benchmarks are started

**Comment by andreafioraldi** [comment]
it is, the afl driver takes input from stdin

**Comment by vanhauser-thc** [comment]
the weird thing is that the aflpp_driver I coded does that fine if I try that locally. it is just the fuzzbench binaries that dont work like that. weird.

**Comment by vanhauser-thc** [comment]
see #404

**Pull request merged**